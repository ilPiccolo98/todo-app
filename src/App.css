td,
th {
  border: black solid 1px;
}

.buttons-panel {
  margin-left: 160px;
}
