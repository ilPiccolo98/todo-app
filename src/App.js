import { BrowserRouter, Link, Switch, Route } from "react-router-dom";
import TableActivities from "./TableActivities/TableActivities";
import AddActivity from "./AddActivity/AddActivity";
import "./App.css";
import { useState } from "react";
import UpdateActivity from "./UpdateActivity/UpdateActivity";
import DeleteActivity from "./DeleteActivity/DeleteActivity";

function App() {
  const [activities, setActivities] = useState([
    {
      id: 1,
      name: "shopping",
      description: "buy some stuff",
      status: false,
    },
    {
      id: 2,
      name: "have fun with my dog",
      description: "take my dog for a walk and have fun with it",
      status: true,
    },
    {
      id: 3,
      name: "work out",
      description: "work out a little bit with weights and barbells",
      status: false,
    },
  ]);
  return (
    <BrowserRouter>
      <TableActivities activities={activities} setActivities={setActivities} />
      <div className="buttons-panel">
        <Link to="/add-activity">
          <input
            data-testid="insert-button-app"
            type="button"
            value="Insert activity"
          />
        </Link>
        <Link to="/update-activity">
          <input
            type="button"
            value="Update activity"
            data-testid="update-button-app"
          />
        </Link>
        <Link to="/delete-activity">
          <input
            data-testid="delete-button-app"
            type="button"
            value="Delete activity"
          />
        </Link>
      </div>
      <Switch>
        <Route
          exact
          path="/add-activity"
          render={() => (
            <AddActivity
              activities={activities}
              setActivities={setActivities}
            />
          )}
        />
        <Route
          exact
          path="/update-activity"
          render={() => (
            <UpdateActivity
              activities={activities}
              setActivities={setActivities}
            />
          )}
        />
        <Route
          exact
          path="/delete-activity"
          render={() => (
            <DeleteActivity
              activities={activities}
              setActivities={setActivities}
            />
          )}
        />
      </Switch>
    </BrowserRouter>
  );
}

export default App;
