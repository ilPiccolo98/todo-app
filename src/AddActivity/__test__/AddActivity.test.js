/* eslint-disable jest/valid-title */
import React from "react";
import App from "../../App";
import { fireEvent, render, waitFor, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

const getAddTaskFormFields = (component) => {
  const nameFieldAddActivity = component.getByTestId("name-field-add-activity");
  const descriptionFieldAddActivity = component.getByTestId(
    "description-field-add-activity"
  );
  const statusCheckboxAddActivity = component.getByTestId(
    "status-checkbox-add-activity"
  );
  const insertButtonAddActivity = component.getByTestId(
    "insert-button-add-activity"
  );

  return {
    nameFieldAddActivity,
    descriptionFieldAddActivity,
    statusCheckboxAddActivity,
    insertButtonAddActivity,
  };
};

const getDeleteActivityFormFields = (component) => {
  const idFieldDeleteActivity = component.getByTestId(
    "id-field-delete-activity"
  );
  const deleteButtonDeleteActivity = component.getByTestId(
    "delete-button-delete-activity"
  );
  return { idFieldDeleteActivity, deleteButtonDeleteActivity };
};

const getRowDataCells = (table, index) => {
  const row = table.childNodes[index];
  const idCellData = row.childNodes[0];
  const nameCellData = row.childNodes[1];
  const descriptionCellData = row.childNodes[2];
  const statusCellData = row.childNodes[3];
  return {
    idCellData,
    nameCellData,
    descriptionCellData,
    statusCellData,
  };
};

describe("AddActivity Component", () => {
  it("should show the empty form after the insert button click", () => {
    const component = render(<App />);
    const insertButtonApp = component.getByTestId("insert-button-app");

    fireEvent.click(insertButtonApp);

    const {
      nameFieldAddActivity,
      descriptionFieldAddActivity,
      statusCheckboxAddActivity,
    } = getAddTaskFormFields(component);

    expect(nameFieldAddActivity.value).toBe("");
    expect(descriptionFieldAddActivity.value).toBe("");
    expect(statusCheckboxAddActivity.checked).toBe(false);
    expect(component).toMatchSnapshot();
  });

  it("should let you type data into the form after pressed insert button", () => {
    const component = render(<App />);
    const insertButtonApp = component.getByTestId("insert-button-app");

    fireEvent.click(insertButtonApp);

    const {
      nameFieldAddActivity,
      descriptionFieldAddActivity,
      statusCheckboxAddActivity,
    } = getAddTaskFormFields(component);

    fireEvent.change(nameFieldAddActivity, {
      target: {
        value: "name new activity",
      },
    });
    fireEvent.change(descriptionFieldAddActivity, {
      target: {
        value: "description new activity",
      },
    });
    fireEvent.click(statusCheckboxAddActivity);
    expect(nameFieldAddActivity.value).toBe("name new activity");
    expect(descriptionFieldAddActivity.value).toBe("description new activity");
    expect(statusCheckboxAddActivity.checked).toBe(true);
  });

  it("should let you insert a new activity", () => {
    const component = render(<App />);
    const insertButtonApp = component.getByTestId("insert-button-app");

    fireEvent.click(insertButtonApp);

    const {
      nameFieldAddActivity,
      descriptionFieldAddActivity,
      statusCheckboxAddActivity,
      insertButtonAddActivity,
    } = getAddTaskFormFields(component);

    fireEvent.change(nameFieldAddActivity, {
      target: {
        value: "name new activity",
      },
    });
    fireEvent.change(descriptionFieldAddActivity, {
      target: {
        value: "description new activity",
      },
    });
    fireEvent.click(statusCheckboxAddActivity);
    fireEvent.click(insertButtonAddActivity);
    const tableBody = component.getByTestId("table-body");
    const { idCellData, nameCellData, descriptionCellData, statusCellData } =
      getRowDataCells(tableBody, tableBody.childElementCount - 1);

    expect(idCellData.textContent).toBe("4");
    expect(nameCellData.textContent).toBe("name new activity");
    expect(descriptionCellData.textContent).toBe("description new activity");
    expect(statusCellData.textContent).toBe("true");
  });

  it("shouldn't let you add an activity with name field blank", () => {
    const component = render(<App />);
    const insertButtonApp = component.getByTestId("insert-button-app");
    fireEvent.click(insertButtonApp);
    const {
      descriptionFieldAddActivity,
      statusCheckboxAddActivity,
      insertButtonAddActivity,
    } = getAddTaskFormFields(component);

    fireEvent.change(descriptionFieldAddActivity, {
      target: {
        value: "description new activity",
      },
    });
    fireEvent.click(statusCheckboxAddActivity);
    fireEvent.click(insertButtonAddActivity);
    const tableBody = component.getByTestId("table-body");
    expect(tableBody.childElementCount).toBe(4);
  });

  it("shouldn't let you add an activity with description field blank", () => {
    const component = render(<App />);
    const insertButtonApp = component.getByTestId("insert-button-app");
    fireEvent.click(insertButtonApp);
    const {
      nameFieldAddActivity,
      statusCheckboxAddActivity,
      insertButtonAddActivity,
    } = getAddTaskFormFields(component);

    fireEvent.change(nameFieldAddActivity, {
      target: {
        value: "name new activity",
      },
    });
    fireEvent.click(statusCheckboxAddActivity);
    fireEvent.click(insertButtonAddActivity);
    const tableBody = component.getByTestId("table-body");
    expect(tableBody.childElementCount).toBe(4);
  });

  it("should delete all the activities and add a new one", async () => {
    const component = render(<App />);
    const deleteButtonApp = component.getByTestId("delete-button-app");
    await waitFor(() => {
      fireEvent.click(deleteButtonApp);
    });
    const { idFieldDeleteActivity, deleteButtonDeleteActivity } =
      getDeleteActivityFormFields(component);

    await waitFor(() => {
      fireEvent.change(idFieldDeleteActivity, {
        target: {
          value: "1",
        },
      });
    });
    await waitFor(() => {
      fireEvent.click(deleteButtonDeleteActivity);
    });
    await waitFor(() => {
      fireEvent.change(idFieldDeleteActivity, {
        target: {
          value: "2",
        },
      });
    });
    await waitFor(() => {
      fireEvent.click(deleteButtonDeleteActivity);
    });
    await waitFor(() => {
      fireEvent.change(idFieldDeleteActivity, {
        target: {
          value: "3",
        },
      });
    });
    await waitFor(() => {
      fireEvent.click(deleteButtonDeleteActivity);
    });

    const insertButtonApp = component.getByTestId("insert-button-app");
    fireEvent.click(insertButtonApp);
    const {
      nameFieldAddActivity,
      descriptionFieldAddActivity,
      statusCheckboxAddActivity,
      insertButtonAddActivity,
    } = getAddTaskFormFields(component);

    fireEvent.change(nameFieldAddActivity, {
      target: {
        value: "name new activity",
      },
    });
    fireEvent.change(descriptionFieldAddActivity, {
      target: {
        value: "description new activity",
      },
    });
    fireEvent.click(statusCheckboxAddActivity);
    fireEvent.click(insertButtonAddActivity);

    const tableBody = component.getByTestId("table-body");

    const { idCellData, nameCellData, descriptionCellData, statusCellData } =
      getRowDataCells(tableBody, 1);

    expect(idCellData.textContent).toBe("1");
    expect(nameCellData.textContent).toBe("name new activity");
    expect(descriptionCellData.textContent).toBe("description new activity");
    expect(statusCellData.textContent).toBe("true");
  });
});
