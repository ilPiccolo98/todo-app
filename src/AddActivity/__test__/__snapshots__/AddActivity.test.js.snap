// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddActivity Component should show the empty form after the insert button click 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <table>
          <tbody
            data-testid="table-body"
          >
            <tr>
              <th>
                Id
              </th>
              <th>
                Name
              </th>
              <th>
                Description
              </th>
              <th>
                Status
              </th>
            </tr>
            <tr>
              <td
                data-testid="td-id-1"
              >
                1
              </td>
              <td
                data-testid="td-name-1"
              >
                shopping
              </td>
              <td
                data-testid="td-description-1"
              >
                buy some stuff
              </td>
              <td
                data-testid="td-status-1"
              >
                false
              </td>
            </tr>
            <tr>
              <td
                data-testid="td-id-2"
              >
                2
              </td>
              <td
                data-testid="td-name-2"
              >
                have fun with my dog
              </td>
              <td
                data-testid="td-description-2"
              >
                take my dog for a walk and have fun with it
              </td>
              <td
                data-testid="td-status-2"
              >
                true
              </td>
            </tr>
            <tr>
              <td
                data-testid="td-id-3"
              >
                3
              </td>
              <td
                data-testid="td-name-3"
              >
                work out
              </td>
              <td
                data-testid="td-description-3"
              >
                work out a little bit with weights and barbells
              </td>
              <td
                data-testid="td-status-3"
              >
                false
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="buttons-panel"
      >
        <a
          href="/add-activity"
        >
          <input
            data-testid="insert-button-app"
            type="button"
            value="Insert activity"
          />
        </a>
        <a
          href="/update-activity"
        >
          <input
            data-testid="update-button-app"
            type="button"
            value="Update activity"
          />
        </a>
        <a
          href="/delete-activity"
        >
          <input
            data-testid="delete-button-app"
            type="button"
            value="Delete activity"
          />
        </a>
      </div>
      <form>
        <div>
          <label
            data-testid="name-label-add-activity"
            for="name"
          >
            Name
          </label>
          <input
            data-testid="name-field-add-activity"
            id="name"
            name="name"
            placeholder="errore"
            value=""
          />
        </div>
        <div>
          <label
            data-testid="description-label-add-activity"
            for="description"
          >
            Description
          </label>
          <input
            data-testid="description-field-add-activity"
            id="description"
            name="description"
            placeholder="Description"
            value=""
          />
        </div>
        <div>
          <label
            data-testid="status-label-add-activity"
            for="status"
          >
            Status
          </label>
          <input
            data-testid="status-checkbox-add-activity"
            id="status"
            name="status"
            type="checkbox"
          />
        </div>
        <div>
          <input
            data-testid="insert-button-add-activity"
            id="submit"
            name="submit"
            type="submit"
            value="Insert"
          />
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <div>
      <table>
        <tbody
          data-testid="table-body"
        >
          <tr>
            <th>
              Id
            </th>
            <th>
              Name
            </th>
            <th>
              Description
            </th>
            <th>
              Status
            </th>
          </tr>
          <tr>
            <td
              data-testid="td-id-1"
            >
              1
            </td>
            <td
              data-testid="td-name-1"
            >
              shopping
            </td>
            <td
              data-testid="td-description-1"
            >
              buy some stuff
            </td>
            <td
              data-testid="td-status-1"
            >
              false
            </td>
          </tr>
          <tr>
            <td
              data-testid="td-id-2"
            >
              2
            </td>
            <td
              data-testid="td-name-2"
            >
              have fun with my dog
            </td>
            <td
              data-testid="td-description-2"
            >
              take my dog for a walk and have fun with it
            </td>
            <td
              data-testid="td-status-2"
            >
              true
            </td>
          </tr>
          <tr>
            <td
              data-testid="td-id-3"
            >
              3
            </td>
            <td
              data-testid="td-name-3"
            >
              work out
            </td>
            <td
              data-testid="td-description-3"
            >
              work out a little bit with weights and barbells
            </td>
            <td
              data-testid="td-status-3"
            >
              false
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="buttons-panel"
    >
      <a
        href="/add-activity"
      >
        <input
          data-testid="insert-button-app"
          type="button"
          value="Insert activity"
        />
      </a>
      <a
        href="/update-activity"
      >
        <input
          data-testid="update-button-app"
          type="button"
          value="Update activity"
        />
      </a>
      <a
        href="/delete-activity"
      >
        <input
          data-testid="delete-button-app"
          type="button"
          value="Delete activity"
        />
      </a>
    </div>
    <form>
      <div>
        <label
          data-testid="name-label-add-activity"
          for="name"
        >
          Name
        </label>
        <input
          data-testid="name-field-add-activity"
          id="name"
          name="name"
          placeholder="errore"
          value=""
        />
      </div>
      <div>
        <label
          data-testid="description-label-add-activity"
          for="description"
        >
          Description
        </label>
        <input
          data-testid="description-field-add-activity"
          id="description"
          name="description"
          placeholder="Description"
          value=""
        />
      </div>
      <div>
        <label
          data-testid="status-label-add-activity"
          for="status"
        >
          Status
        </label>
        <input
          data-testid="status-checkbox-add-activity"
          id="status"
          name="status"
          type="checkbox"
        />
      </div>
      <div>
        <input
          data-testid="insert-button-add-activity"
          id="submit"
          name="submit"
          type="submit"
          value="Insert"
        />
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
